scenario/state.py

```diff
@@ -1760,8 +1774,86 @@ def _is_builtin_event(self, charm_spec: "_CharmSpec"):
+            )
+        # Walk the module looking for the class.
+        for charm_class in ast.walk(tree):
+            if (
+                isinstance(charm_class, ast.ClassDef)
+                and charm_class.name == charm_type.__name__
+            ):
+                break
+        else:
+            raise ValueError(
+                "cannot infer the handler (can't find class) - please pass it explicitly.",
+            )
+        # Walk the class looking for __init__.
+        for init_method in ast.walk(charm_class):
+            if (
+                isinstance(init_method, ast.FunctionDef)
+                and init_method.name == "__init__"
+            ):
+                break
+        else:
+            raise ValueError(
+                "cannot infer the handler (can't find __init__) - please pass it explicitly.",
+            )
+        # Walk __init__ looking for calls to a method named "observe". It's possible that there
+        # is some other method that's not framework.observe, but for our purposes this should be
+        # accurate enough. In cases where we can't find the correct method, the test should
+        # explicitly pass the handler. The same applies if the observe is outside of the __init__
+        # method of the class (in a lib, for example), rather than recursively walking through all
+        # of the __init__ code.
+        handler = None
+        for node in ast.walk(init_method):
+            if (
+                not isinstance(node, ast.Call)
+                or not isinstance(node.func, ast.Attribute)
+                or node.func.attr != "observe"
+                or not node.args
+                or len(node.args) != 2
+            ):
+                continue
+            event, event_handler = node.args
+            if not isinstance(event, ast.Attribute) or not isinstance(
+                event_handler,
+                ast.Attribute,
+            ):
+                # There may be other possibilities, but stick with the most common case for now.
+                continue
+            if event.attr == self.name:
+                if handler is not None:
+                    raise ValueError(
+                        "cannot infer the handler (multiple observers) - please pass it explicitly",
+                    )
+                handler = event_handler.attr
+        if handler is None:
+            raise ValueError(
+                "cannot infer the handler (can't find any observer) - please pass it explicitly",
+            )
+        return getattr(charm_type, handler)
+
+    def deferred(
+        self,
+        handler: Optional[Callable] = None,
+        event_id: int = 1,
+    ) -> DeferredEvent:
         """Construct a DeferredEvent from this Event."""
+        if handler is None:
+            if self.context is None:
+                raise ValueError(
+                    "Handler must be provided if the event is not bound to a Context.",
```

Comment:
Yeah, it shouldn't happen, because you're (in 7.0) meant to only get `_Event`s via the `ctx.on` mechanism rather than ever manually constructing them, and all of those make sure that there is a `context`. It should be required rather than optional, I guess, which would avoid this. But also plays into the discussion in a different thread about whether it needs to be there all the time or not.